[newrelic-infra-operator] Add chart for the New Relic Infra Operator (#347)
newrelic/helm-charts: charts/newrelic-infra-operator/README.md

| `config.infraAgentInjection.agentConfig.configSelectors[].labelSelector`        | LabelSelector matching the labels of the mutating pods.                                                                                                                                                                                                                                                                                     |                                                                                                      |

## Example

Make sure you have [added the New Relic chart repository.](../../README.md#install)

Then, to install this chart, run the following command:

```sh

> helm upgrade --install newrelic/newrelic-infra-operator --set cluster=my_cluster_name

Reviewer: License is also required, right?

Reviewer 2: yes I am adding it, it was also missing from the readme

Reviewer 2: as well as "cluster"

Reviewer 2: also missing release name from the command

```

## Configure in which pods the sidecar should be injected

Policies are available in order to configure in which pods the sidecar should be injected.
Each policy is evaluated independently and if at least one policy matches the operator will inject the sidecar.

Policies are composed by `namespaceSelector` checking the labels of the Pod namespace, `podSelector` checking
the labels of the Pod and `namespace` checking the namespace name. Each of those, if specified, are ANDed.

> By default, the policies are configured in order to inject the sidecar in each pod belonging to a fargate profile.

Reviewer: ```suggestion
By default, the policies are configured in order to inject the sidecar in each pod belonging to a Fargate profile.
```

>Moreover, it is possible to add the label `infra-operator.newrelic.com/disable-injection` to Pods to exclude injection

> for single Pods that otherwise would be selected by the policies.

Reviewer: ```suggestion
for a single Pod that otherwise would be selected by the policies.
```

> Please make sure to configure policies correctly avoiding injecting sidecar for pods running on EC2 nodes

Reviewer: ```suggestion
Please make sure to configure policies correctly to avoid injecting sidecar for pods running on EC2 nodes
```

already monitored by the infrastructure DaemonSet.

## Configure the sidecar with labelsSelectors

It is also possible to configure `resourceRequirements` and `extraEnvVars` based on the labels of the mutating Pod.

The current configuration increases the resource requirements for sidecar injected on `KSM` instances. Moreover, 

> injectes disable the `DISABLE_KUBE_STATE_METRICS` environmentVariable for pods not running on `KSM` instances

Reviewer: ```suggestion
injectes disable the `DISABLE_KUBE_STATE_METRICS` environment variable for Pods not running on `KSM` instances
```

> to decrease the load on the ApiServer.

Reviewer: ```suggestion
to decrease the load on the API server.
```

# newrelic-infra-operator

## Chart Details

This chart will deploy the [New Relic Infrastructure Operator][1], which injects the New Relic Infrastructure solution as a sidecar to specific pods. This is typically used in environments where DaemonSets are not available, such as EKS Fargate.

## Configuration

| Parameter                     | Description                                                  | Default                    |
| ----------------------------- | ------------------------------------------------------------ | -------------------------- |

> | `cluster`                     | The cluster name for the Kubernetes cluster.                 |                            |

Reviewer: We should document other parameters later.

> TBD

## Example

Make sure you have [added the New Relic chart repository.](../../README.md#installing-charts)

Then, to install this chart, run the following command:

```sh
helm upgrade --install newrelic/newrelic-infra-operator --set cluster=my_cluster_name
```

## Resources

The default set of resources assigned to the pods is shown below:

>     resources:
>       limits:
>         memory: 80M
>       requests:
>         cpu: 100m
>         memory: 30M

Reviewer: ```suggestion
    ```yaml
    resources:
      limits:
        memory: 80M
      requests:
        cpu: 100m
        memory: 30M
    ```
```

> # newrelic-metadata-injection

Reviewer: Needs updating.

## Chart Details

> This chart will deploy the [New Relic Infrastructure metadata injection webhook][1].

Reviewer: Needs updating.

## Configuration

| Parameter                     | Description                                                  | Default                    |
| ----------------------------- | ------------------------------------------------------------ | -------------------------- |
| `cluster`                     | The cluster name for the Kubernetes cluster.                 |                            |
| `injectOnlyLabeledNamespaces` | Limit the injection of metadata only to specific namespaces that match the label `newrelic-metadata-injection: enabled`. | false |

> | `image.repository`            | The container to pull.                                       | `newrelic/k8s-metadata-injection`   |

Reviewer: I suggest we align table using http://markdowntable.com/.

| `podSecurityContext.runAsGroup`| runAsUser GID for Pod Security Context                      | `1001`                              |
| `podAnnotations`              | If you wish to provide additional annotations to apply to the pod(s), specify them here.      |                                 |
| `priorityClassName`           | Scheduling priority of the pod                               | `nil`                               |
| `nodeSelector`                | Node label to use for scheduling                             | `{}`                                |
| `timeoutSeconds`              | Seconds to wait for a webhook to respond. The timeout value must be between 1 and 30 seconds| `10`                             |
| `tolerations`                 | List of node taints to tolerate (requires Kubernetes >= 1.6) | `[]`                                |
| `affinity`                    | Node affinity to use for scheduling                          | `{}`                                |

## Example

Make sure you have [added the New Relic chart repository.](../../README.md#installing-charts)

Then, to install this chart, run the following command:

```sh

> helm install newrelic/newrelic-mutation-webhook --set cluster=my_cluster_name

Reviewer: Needs updating.

```

## Resources

The default set of resources assigned to the pods is shown below:

    resources:
      limits:
        memory: 80M
      requests:
        cpu: 100m
        memory: 30M

> [1]: https://docs.newrelic.com/docs/integrations/kubernetes-integration/link-your-applications/link-your-applications-kubernetes#configure-injection

Reviewer: This can be removed I think?

## Configuration

| Parameter                     | Description                                                  | Default                    |
| ----------------------------- | ------------------------------------------------------------ | -------------------------- |
| `cluster`                     | The cluster name for the Kubernetes cluster.                 |                            |

> | `image.repository`            | The container to pull.                                       | `newrelic/newrelic-infra-operator`   |

Reviewer: I'd fix indentation of table with http://markdowntable.com/ :)

| `tolerations`                                                                   | List of node taints to tolerate (requires Kubernetes >= 1.6)                                                                                                                                                                                                                                                                                | `[]`                                                                                                 |
| `affinity`                                                                      | Node affinity to use for scheduling.                                                                                                                                                                                                                                                                                                         | `{}`                                                                                                 |

> | `config.ignoreMutationErrors`                                                   | if true it instruments the operator to ignore injection error instead of failing.                                                                                                                                                                                                                                                            | `true`                                                                                               |

Reviewer: I wonder if this is the right place to document the syntax of the configuration. Maybe it would be better to document it in operator docs instead?

| `config.infraAgentInjection.policies[].podSelector`                             | Selector on Pod Labels.                                                                                                                                                                                                                                                                                                                     |                                                                                                      |
| `config.infraAgentInjection.policies[].namespaceSelector`                       | Selector on Namespace labels.                                                                                                                                                                                                                                                                                                               |                                                                                                      |

> | `config.infraAgentInjection.policies[].namespace`                               | Namespace name. if set only pods belonging to such namespace matches the policy.                                                                                                                                                                                                                                                             |                                                                                                      |

Reviewer: ```suggestion
| `config.infraAgentInjection.policies[].namespace`                               | Namespace name. If set only pods belonging to such namespace matches the policy.                                                                                                                                                                                                                                                             |                                                                                                      |
```

> | `config.infraAgentInjection.customAttributes[].name`                            | name of custom attribute to include.                                                                                                                                                                                                                                                                                                         |                                                                                                      |

Reviewer: And all descriptions below should be capitalized as well.
```suggestion
| `config.infraAgentInjection.customAttributes[].name`                            | Name of custom attribute to include.                                                                                                                                                                                                                                                                                                         |                                                                                                      |
```

| `config.infraAgentInjection.agentConfig.image.repository`                       | The infrastructure agent repository for the sidecar container.                                                                                                                                                                                                                                                                              | `newrelic/infrastructure-k8s`                                                                        |
| `config.infraAgentInjection.agentConfig.image.pullPolicy`                       | The sidecar image pull policy.                                                                                                                                                                                                                                                                                                              | `IfNotPresent`                                                                                       |

> | `config.infraAgentInjection.agentConfig.image.tag`                              | The infrastructure agent repository for the sidecar container.                                                                                                                                                                                                                                                                              | `2.4.0-unprivileged`                                                                                 |

Reviewer: Perhaps tag should be documented after repository? It makes sense from both logical point of view and from alphabetical order.
```suggestion
| `config.infraAgentInjection.agentConfig.image.tag`                              | The infrastructure agent image tag for the sidecar container.                                                                                                                                                                                                                                                                              | `2.4.0-unprivileged`                                                                                 |
```

Reviewer 2: 👍

| `config.infraAgentInjection.agentConfig.image.repository`                       | The infrastructure agent repository for the sidecar container.                                                                                                                                                                                                                                                                              | `newrelic/infrastructure-k8s`                                                                        |

> | `config.infraAgentInjection.agentConfig.podSecurityContext.RunAsUser`           | runAsUser UID for Pod Security Context.                                                                                                                                                                                                                                                                                                      |                                                                                                      |

Reviewer: I don't think field should be capitalized.
```suggestion
| `config.infraAgentInjection.agentConfig.podSecurityContext.runAsUser`           | runAsUser UID for Pod Security Context.                                                                                                                                                                                                                                                                                                      |                                                                                                      |
```
